### PR TITLE
config variable to raise UserWarning if model not fully defined

### LINKED ISSUE #4683: V2: config parameter to warning about missing annotations
We have [logic](https://github.com/pydantic/pydantic/blob/eaa2ad14c71138600de301d5f0ebf8fda6775788/pydantic/_internal/_model_construction.py#L203) to deal with the case where an annotation is not defined when a model is constructed, at the moment this silently stops the schema/validator build and `__pydantic_model_complete__` is left `False` in the hope that the missing annotation(s) can be defined and the model completed via `model_rebuild`.

I think we should have a config flag, default True which raising a warning when this happens - it should be a `UserWarning` so it gets printed by default.

Needs #4516 to be merged before this can be worked on, but would be amazing if someone could take this on.

### PR DESCRIPTION
**Pull request by jimkring**
Raises a UserWarning if model not fully defined and will need rebuild.

fix #4683 -> **V2: config parameter to warning about missing annotations**

New config item added to BaseConfig:
`warn_on_undefined_types: bool = True`

Also added small block of code to `build_inner_schema()` that checks if any of the annotations are of type`ForwardRef` and raises a PydanticUndefinedAnnotation error, in the same way as would happen in the missing annotation were caught as an AttributeError when using `from __future__ import annotations` instead of a `ForwardRef`

### CODE REVIEW

**Comment by jimkring** [comment]
@samuelcolvin I can't seem to get CI to pass since pre-commit wants to Pyupgrade `foo: Optional[Foo] = None` to `foo: Foo | None = None`, but that causes CI tests on older python versions to fail. Not sure best path forward.

File: pydantic/config.py
Comment (ybressler) [inline pydantic/config.py:86]: It's technically not an undefined type, rather, a non-annotated type. I guess this is pedantics (which is the whole point 😉 ), but we should choose a suitable name for this config arg.

File: tests/test_forward_ref.py
Comment (ybressler) [inline tests/test_forward_ref.py:49]: This is reused many times in the test. Perhaps create a new BaseClass with the proper config and inherit? (Could make tests dry-er)

File: pydantic/_internal/_model_construction.py
Comment (ybressler) [inline pydantic/_internal/_model_construction.py:143]: Vestige of development, remove:
```suggested
        warning_string = f'`{name}` is not fully defined, you should define `{e}`, then call `{name}.model_rebuild()`'
```

File: pydantic/_internal/_model_construction.py
Comment (ybressler) [inline pydantic/_internal/_model_construction.py:210]: Is this necessary? Looks erroneous?

File: pydantic/config.py
Comment (ybressler) [inline pydantic/config.py:86]: Also, is there documentation which needs updating? (Readme's etc.?)

File: tests/test_construction_missing_annotations.py
Comment (ybressler) [inline tests/test_construction_missing_annotations.py:20]: Which line exactly is expected to raise the error? Is it line 17? Or line 20? (Perhaps indicate so in the docstring of the test.)

File: tests/test_construction_missing_annotations.py
Comment (ybressler) [inline tests/test_construction_missing_annotations.py:42]: This test should be revised to be a bit simpler. I gather that you are providing a reference in class `Foo` to `Bar`, the latter of which doesn't have an annotation. Thereafter, you provide `Bar` an annotation and expect to rebuild `Foo` without error. If yes, then building `Bar` to include `Foo` is unnecessarily complex. Why not provide `Bar` with simpler attributes, since this test just needs to validate that `Foo.__pydantic_model_complete__ is True` (Bar can be ignored...)

File: tests/test_construction_missing_annotations.py
Comment (ybressler) [inline tests/test_construction_missing_annotations.py:45]: Function name should be a bit more specific IMO.

File: tests/test_construction_missing_annotations.py
Comment (ybressler) [inline tests/test_construction_missing_annotations.py:56]: This test looks incomplete.

**Review by ybressler — CHANGES_REQUESTED** [review]
Great work thus far. Left some feedback + some Q's for you.

**Comment by ybressler** [comment]
> @samuelcolvin I can't seem to get CI to pass since pre-commit wants to Pyupgrade `foo: Optional[Foo] = None` to `foo: Foo | None = None`, but that causes CI tests on older python versions to fail. Not sure best path forward.

A simpler (more focused) test case would solve this issue for you IMO.

File: pydantic/_internal/_model_construction.py
Comment (samuelcolvin) [reply pydantic/_internal/_model_construction.py:143]: why?

File: pydantic/config.py
Comment (samuelcolvin) [reply pydantic/config.py:86]: I think "undefined" is the best term. This warning related to annotations which are not yet defined when creating the model.

This is not about non-annotated types.

Comment (samuelcolvin) [reply pydantic/config.py:86]: yes, would be useful to add documentation for this now. Although the docs needs a lot of work for V2. 

@jimkring Please could you and something to the docs, and I'll take care of migrating it to the new documentation structure when the time comes.

File: tests/test_construction_missing_annotations.py
Comment (samuelcolvin) [inline tests/test_construction_missing_annotations.py:1]: I think these tests can go in `test_forward_ref.py`.

File: tests/test_construction_missing_annotations.py
Comment (samuelcolvin) [inline tests/test_construction_missing_annotations.py:14]: Better to use `create_module` and `from __future__ import annotations`, then just have

```py
from __future__ import annotations
from pydantic import BaseModel

class Foobar(BaseModel):
    a: MissingType
```

Comment (samuelcolvin) [reply tests/test_construction_missing_annotations.py:14]: you can add this self-referencing case as a separate test. Best to match against more of the error message to confirm where the error is happening.

File: pydantic/_internal/_model_construction.py
Comment (samuelcolvin) [inline pydantic/_internal/_model_construction.py:208]: This makes sense, but I don't think it needs to be a separate loop.

File: pydantic/_internal/_model_construction.py
Comment (samuelcolvin) [reply pydantic/_internal/_model_construction.py:210]: See the PR description.

I think this case is tested, @jimkring please confirm.

File: pydantic/config.py
Comment (samuelcolvin) [reply pydantic/config.py:86]: `undefined_types_warning` would be most concise.

**Review by samuelcolvin — COMMENTED** [review]
This is looking great, a few more things to fix.

Please update.

File: pydantic/config.py
Comment (ybressler) [reply pydantic/config.py:86]: Thanks for the clarification.

File: pydantic/_internal/_model_construction.py
Comment (ybressler) [reply pydantic/_internal/_model_construction.py:143]: ^ I attempted to leave a code suggestion, but it didn't go. What my comment **meant** to say is >> "remove the print statement"

File: tests/test_construction_missing_annotations.py
Comment (ybressler) [reply tests/test_construction_missing_annotations.py:42]: Curious if you feel strongly 1 way or the other @samuelcolvin

File: pydantic/_internal/_model_construction.py
Comment (jimkring) [reply pydantic/_internal/_model_construction.py:209]: Oops, I deleted my old reply/comment somehow. The point is that I think it does need to be a separate loop, since there is really no "other loop" to put this in (that I can find).

**Review by jimkring — COMMENTED** [review]
All CI tests are passing and I think I'm "done" with this round of edits. Look forward to your review.

File: tests/test_forward_ref.py
Comment (jimkring) [inline tests/test_forward_ref.py:876]: @samuelcolvin this is a reusable assertion that requires a block of code to raise an "undefined types warning". It requires matching the exact warning string, including the specific class names involved.

File: tests/test_forward_ref.py
Comment (jimkring) [inline tests/test_forward_ref.py:953]: @samuelcolvin @ybressler there are now 4 new unit tests which are basically a matrix of:

1) how to handle UndefinedType (import annotations vs import ForwardRef)
    a) use `from __future__ import annotations`
    b) use `UndefinedType = ForwardRef('UndefinedType')`
2) whether to suppress the undefined_types_warning (suppress warning or not)
    a) don't suppress it and then verify the expected warning and its detailed message
    b) do suppress it and then verify the model is partially built (without the warning being raised)

The test names reflect these 4 different possibilities of the two different parameters

File: tests/test_forward_ref.py
Comment (jimkring) [inline tests/test_forward_ref.py:50]: @samuelcolvin @ybressler there are several pre-existing tests that needed `undefined_types_warning = False` added to their config, in order for them to work as they did previously (since the previous behavior when using a `ForwardRef` was a silent fail instead of an undefined_types_warning).  We should make sure this makes sense for all the affected tests. You both (and/or the previous test authors) are more qualified (than I am) to make that determination.

Comment (samuelcolvin) [reply tests/test_forward_ref.py:50]: Makes sense, I'll look.

File: tests/test_forward_ref.py
Comment (samuelcolvin) [reply tests/test_forward_ref.py:953]: Great.

File: tests/test_forward_ref.py
Comment (ybressler) [inline tests/test_forward_ref.py:892]: Fancy stuff. I like it!

File: tests/test_forward_ref.py
Comment (ybressler) [inline tests/test_forward_ref.py:930]: Nice test. Looks very similar to the test `test_undefined_type_raises_warning()`. There are 2 general ways to make life simpler here:
1. Describe the relationship of these 2 tests in their docstrings. (Human talking to a human solution)
2. Parameterize these into a single test. In this case, you get tests which is slightly DRY-er, but slightly more more complex.

I personally am in favor of the 2nd option. Btw, feel free to delegate to a follow-up PR which I would be happy to contribute to – OR, I can contribute to this PR if you want support with this. (OR, choose option 1.) _OR..._ something else?

File: tests/test_forward_ref.py
Comment (ybressler) [inline tests/test_forward_ref.py:933]: Same feedback as above re: parameterizing tests (related to `test_undefined_type_raises_warning_if_forward_ref()`

File: tests/test_forward_ref.py
Comment (ybressler) [reply tests/test_forward_ref.py:953]: Fantastic work!

File: pydantic/_internal/_model_construction.py
Comment (ybressler) [inline pydantic/_internal/_model_construction.py:211]: I don't see a test which correctly raises this error.

In other words, if I were to check out your branch and remove this code (or change it to raise no error), I would want a test to fail.

File: docs/examples/types_undefined_warning.py
Comment (ybressler) [inline docs/examples/types_undefined_warning.py:50]: Instead of `jane_doe.json()` I'd rather `jane_doe.dict()`, allows neater evaluation of python objects.

**Review by ybressler — CHANGES_REQUESTED** [review]
Thanks for your quick turnarounds! Some additional comments / feedback for you. I think we're almost done.

File: pydantic/_internal/_model_construction.py
Comment (jimkring) [reply pydantic/_internal/_model_construction.py:211]: @ybressler when I comment out this block on my dev system (Python 3.10), the two tests fail:

- test_undefined_type_raises_warning_if_forward_ref
- test_disable_undefined_types_warning_if_missing_class_is_a_forwardref

This was the reason I added that block of code.  This makes me wonder if there are different behaviors on different Python versions or systems, since the importance of this block of code was initially questioned by you and @samuelcolvin.

Ah, yes... I think I found the source of the problem/difference:

Inside `_typing_extra.get_type_hints()` it checks if Python >= 3.10 and then uses:

Python >= 3.10 -> the built-in `typing.get_type_hints()` for Python >= 3.10
Python < 3.10 -> a special implementation with partially copied code from 3.10

_typing_extra.py lines 216-234
```python
if sys.version_info >= (3, 10):  # noqa C901
    get_type_hints = typing.get_type_hints

else:
    """
    For older versions of python, we have a custom implementation of `get_type_hints` which is a close as possible to
    the implementation in CPython 3.10.8.
    """
    fr_has_is_class = True

    def ForwardRefWrapper(arg: Any, is_argument: bool = True, *, is_class: bool = False) -> typing.ForwardRef:
        """
        Wrapper for ForwardRef that accounts for the `is_class` argument missing in older versions.
        The `module` argument is omitted as it breaks <3.9 and isn't used in the calls below.

        See https://github.com/python/cpython/pull/28560 for some background

        Implemented as EAFP with memory.
        """
```

Untangling this is probably important and is something I will defer to others :)

It's great that you noticed this @ybressler. My guess is that this will make sense to @samuelcolvin since the custom implementation of `get_type_hints` appears to be his work.

File: tests/test_forward_ref.py
Comment (jimkring) [reply tests/test_forward_ref.py:930]: @ybressler I see what you mean, yet I'm not quite sure how to go about parameterizing a test that use `create_module`. Any code defined outside the `def module()` function is easily accessible inside the `module` that is the code under test.  Also, for tests where the `module` is calling `from __future__ import annotations` we need to `create_module()` from a string `# language=Python` otherwise it seems the tests don't run correctly (I'm not entirely sure why, but generally `from __future__ import annotations` has the be the first line of code in a file).  So, for the sake of symmetry and readability, I'm inclined the just sprinkle in some additional comments (1st option you recommended).  All that said, I'm very curious what changes you'd propose (since I'd probably learn a lot and you're probably able to see further around the obstacles than I can) -- I would probably defer that to a follow-up PR, once the smoke settles around the questions of ForwardRef handling (in Python < 3.10 vs. >=3.10).

File: pydantic/_internal/_model_construction.py
Comment (samuelcolvin) [reply pydantic/_internal/_model_construction.py:209]: line 214 we iterate over `type_hints`

```py
for ann_name, ann_type in type_hints.items():
```

Comment (jimkring) [reply pydantic/_internal/_model_construction.py]: It was hiding in plain sight :)
Thanks!

Comment (jimkring) [reply pydantic/_internal/_model_construction.py]: fixed and pushed

File: docs/examples/types_undefined_warning.py
Comment (jimkring) [reply docs/examples/types_undefined_warning.py]: Nice suggestion! I made the change (to what's shown below) and pushed:

```python
assert jane_doe.dict(exclude_unset=True) == {
    'name': 'Jane Doe',
    'books_read': [
        {
            'title': 'Python Crash Course',
            'author': {'name': 'Eric Matthes'},
        },
    ],
}
```

File: tests/test_forward_ref.py
Comment (jimkring) [reply tests/test_forward_ref.py:892]: Thanks!

File: tests/test_forward_ref.py
Comment (ybressler) [reply tests/test_forward_ref.py:930]: Good approach. Once comments are added, let's close this thread.

File: pydantic/_internal/_model_construction.py
Comment (ybressler) [reply pydantic/_internal/_model_construction.py]: Agreed. Let's wait till Samuel reviews. I imagine a suitable approach is to create a follow up issue (perhaps backlogged) to untangle this functionality. Meanwhile, code is good as is.

File: docs/examples/types_undefined_warning.py
Comment (ybressler) [inline docs/examples/types_undefined_warning.py:41]: As long as there's one book, you won't have an issue. But if there are more books, you may get into an issue with ordering. That's not for now though. I think we're good to solve that problem later, should it present itself.

**Review by ybressler — APPROVED** [review]
Looks good to me! Let's resolve that one last thread about `python3.10` with `type_hints` before calling this done.

Great work, and very quick too!

File: tests/test_forward_ref.py
Comment (ybressler) [inline tests/test_forward_ref.py:888]: Probably the best inline documentation I've seen in a very long time. Well done.

**Review by ybressler — APPROVED** [review]

**Comment by samuelcolvin** [comment]
I'll try and review tomorrow.

File: tests/test_forward_ref.py
Comment (jimkring) [inline tests/test_forward_ref.py:957]: @ybressler I've added a block of comments/docs at the top of these 4 tests to describe:

- why the tests have duplicated code (and some context to help someone untangle it the future)
- what the tests are doing (with respect to the parameters/conditions we're trying to cover).
- some simple tags for each parameter state that can be used in the test name (and provide a reference to the comment block in order to to avoid duplication of the notes/comments within the tests themselves).

File: tests/test_forward_ref.py
Comment (jimkring) [reply tests/test_forward_ref.py]: comments added. I'll close this thread and we can review the changes in that other thread.

File: tests/test_forward_ref.py
Comment (jimkring) [reply tests/test_forward_ref.py:888]: thanks! I'll mark this as resolved

File: tests/test_forward_ref.py
Comment (ybressler) [reply tests/test_forward_ref.py:957]: Excellent!

**Comment by jimkring** [comment]
Thanks @samuelcolvin. I'm happy to discuss if you have any questions.

**Comment by samuelcolvin** [comment]
Sorry, been busy, I'll review soon.

**Comment by samuelcolvin** [comment]
Sorry, been busy, I'll review soon.

File: docs/examples/types_undefined_warning.py
Comment (samuelcolvin) [inline docs/examples/types_undefined_warning.py:26]:
```suggestion
# Now, we can rebuild the `Book` model, since the `Person` model is now defined.
# Note: there's no need to call `model_rebuild()` on `Person`, it's already complete.
Book.model_rebuild()
```

**Review by samuelcolvin — APPROVED** [review]
otherwise LGTM. I'll add that suggestion and merge.

**Comment by samuelcolvin** [comment]
Thanks so much @jimkring, this is great.

File: docs/examples/types_undefined_warning.py
Comment (samuelcolvin) [inline docs/examples/types_undefined_warning.py:26]:
```suggestion
# Note: there's no need to call `model_rebuild()` on `Person`, 
# it's already complete.
```

**Comment by jimkring** [comment]
Thanks @samuelcolvin.
@ybressler: I just fixed a minor linting error (EOL) to get CI into a passing state.  It should be ready to merge.

Thank you both for taking the time to do the code reviews and offer suggestions. I learned a lot in the process and it feels great to support this impactful project.

**Comment by samuelcolvin** [comment]
Thanks so much, merged.

**Pull request merged**
